### DISCUSSION TITLE
Nesting Behavior Trees

### DISCUSSION THREAD

**Discussion by Agaanii** [bitbrain/beehave#238] [Ideas]
I'm working on an RTS project at the moment, and I'm testing out Beehave for my unit behaviors.

A challenge I'm running into is, I'd like to be able to compose behavior trees of behavior trees. For example, two useful commands in the RTS are Attack-Move and Pursue-And-Attack-Target.

Attack-Move commands should be roughly:
* If you have a target
** If they're in range, attack them
** If they're not in attack range, but they're still nearby, pursue them
* Otherwise, see if there's a potential target in range.
** If so, assign them as target. Go to next tick
* Otherwise, move toward next waypoint

Pursue-and-attack-target is somewhat different
* If target is in range, attack them
* If target is not in range, pursue them

There are places there where I'd love to be able to define a couple common behaviors, like "Repeatedly attack target that is in range" and "Pursue target that is out of range."

To that end, I'd like to be able to have an Action node that is "Run child behavior tree" so that I can build those simpler behaviors and compose them into more complex behaviors. It would also support the general RTS design of having a set of commands you give to units, who then set those commands as their currently-designated behavior and run them.

Perhaps I'm missing something and this is already possible?

**Comment by lostptr** [comment]
The fact that behavior trees are nodes already allows us to do this.
You can just create a scene like this:
- AttackMove (reactive selector node)
  - On target found (sequence node)
    - Has target (condition node)
    - Attack if in range (sequence node)
      - Is in range (condition node)
      - Attack (action node)
    - Pursue (action node)
  - Try finding new target (action node)
  - Move to point (action node)

And now you can instanciate the AttackMove scene inside any beehave tree!

**Comment by Agaanii** [comment]
Thanks! I hadn't thought to have a Selector/Sequence node as a scene root, that makes total sense and will support my needs!
